File: Repository/src/main/java/example/repo/OfficeDBRepository.java
package example.repo;

import Utils.JdbcUtils;
import example.model.Event;
import example.model.Office;
import example.model.Participant;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class OfficeDBRepository implements OfficeRepository {
    private final JdbcUtils dbUtils;
    private ParticipantRepository participantRepository;
    private EventRepository eventRepository;
    private static final Logger logger = LogManager.getLogger();

    public OfficeDBRepository(Properties props, ParticipantRepository participantRepository, EventRepository eventRepository) {
        logger.info("Initializing OfficeDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
        this.participantRepository = participantRepository;
        this.eventRepository = eventRepository;
    }

    @Override
    public Collection<Participant> findParticipantsByEvent(Long eventId) {
        String sql = "SELECT idParticipant, name, age FROM Offices WHERE idEvent = ?";
        Connection connection = dbUtils.getConnection();
        List<Participant> participants = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setLong(1, eventId);
            ResultSet rs = preparedStatement.executeQuery();
            while (rs.next()) {
                Long id = rs.getLong("idParticipant");
                String name = rs.getString("name");
                int age = rs.getInt("age");
                Participant participant = new Participant(name, age);
                participant.setId(id);
                participants.add(participant);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return participants;
    }

    @Override
    public int countEventsForParticipant(Long participantId) throws EntityRepoException {
        String sql = "SELECT COUNT(*) FROM Offices WHERE idParticipant = ?";
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setLong(1, participantId);
            ResultSet resultSet = preparedStatement.executeQuery();

            return resultSet.next() ? resultSet.getInt(1) : 0;
        } catch (SQLException e) {
            throw new EntityRepoException(e);
        }
    }


    @Override
    public void add(Office entity) throws EntityRepoException {
        logger.traceEntry("Adding office {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "INSERT INTO Offices(idEvent, style, distance, idParticipant, name, age) VALUES (?,?,?,?,?,?)";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, entity.getEvent().getId());
            ps.setString(2, entity.getEvent().getStyle());
            ps.setInt(3, entity.getEvent().getDistance());
            ps.setLong(4, entity.getParticipant().getId());
            ps.setString(5, entity.getParticipant().getName());
            ps.setInt(6, entity.getParticipant().getAge());
            ps.executeUpdate();
            logger.traceExit("Office {} added", entity);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public void remove(long id) throws EntityRepoException {
        logger.traceEntry("Removing office {} ", id);
        Connection connection = dbUtils.getConnection();
        String sql = "DELETE FROM Offices WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
            logger.traceExit("Office {} removed", id);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public List<Office> getAll() throws EntityRepoException {
        logger.traceEntry("Fetching all offices");
        Connection connection = dbUtils.getConnection();
        List<Office> offices = new ArrayList<>();
        String sql = "SELECT * FROM Offices";
        try (PreparedStatement ps = connection.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                Office office = extract(rs);
                offices.add(office);
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        return offices;
    }

    @Override
    public Office findById(long id) throws EntityRepoException {
        logger.traceEntry("Finding office by ID {}", id);
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Offices WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    Office office = extract(rs);
                    logger.traceExit("Found office {}", office);
                    return office;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new EntityRepoException(e);
        }
        return null;
    }

    @Override
    public void update(long id, Office entity) throws EntityRepoException {
        logger.traceEntry("Updating office {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "UPDATE Offices SET idEvent=?, style=?, distance=?, idParticipant=?, name=?, age=? WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, entity.getEvent().getId());
            ps.setString(2, entity.getEvent().getStyle());
            ps.setInt(3, entity.getEvent().getDistance());
            ps.setLong(4, entity.getParticipant().getId());
            ps.setString(5, entity.getParticipant().getName());
            ps.setInt(6, entity.getParticipant().getAge());
            ps.setLong(7, id);
            ps.executeUpdate();
            logger.traceExit("Office {} updated", entity);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    @Override
    public Collection<Office> getEntriesByEvent(Long eventID) {
        logger.traceEntry("getEntriesByEvent with task od id={} ", eventID);
        Connection connection = dbUtils.getConnection();
        Collection<Office> offices = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement(
                "SELECT * FROM Offices WHERE idEvent=?")) {

            ps.setLong(1, eventID);

            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    Office office = extract(rs);
                    offices.add(office);
                }
            }
            logger.traceExit("Found {} offices for event {}", eventID);
        } catch (SQLException | EntityRepoException e) {
            logger.error(e);
            System.err.println("Error getting offices by event: " + e.getMessage());
        }
        return offices;
    }

    @Override
    public void deleteByIDs(Long participantID, Long eventID) {
        logger.traceEntry("deleting task of participantID={}, eventID={} ", participantID, eventID);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "Delete from Offices where idParticipant=? and idEvent=?")) {
            preparedStatement.setLong(1, participantID);
            preparedStatement.setLong(2, eventID);
            preparedStatement.executeUpdate();
            logger.traceExit("Offices {} deleted", participantID);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error deleting task of participantID=" + participantID);
        }
    }

    private Office extract(ResultSet resultSet) throws SQLException, EntityRepoException {
        Long eventId = resultSet.getLong("idEvent");
        String eventStyle = resultSet.getString("style");
        Integer eventDistance = resultSet.getInt("distance");
        Event event = new Event(eventStyle, eventDistance);
        event.setId(eventId);
        Long participantId = resultSet.getLong("idParticipant");
        String participantName = resultSet.getString("name");
        Integer participantAge = resultSet.getInt("age");
        Participant participant = new Participant(participantName, participantAge);
        participant.setId(participantId);
        Office entry = new Office(participant, event);
        return entry;
    }

}

File: java-app/csharp-app/Repository/OfficeDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using mpp_proiect_csharp_DianaGliga11.Model;
using mpp_proiect_csharp_DianaGliga11.Repository;

namespace mpp_proiect_csharp_DianaGliga11.Repository
{
    public class OfficeDBRepository : I_OfficeDBRepository
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(OfficeDBRepository));
        private readonly IDictionary<string, string> Props;
        private readonly I_ParticipantDBRepository participantRepository;
        private readonly I_EventDBRepository eventRepository;
        
        public OfficeDBRepository(IDictionary<string, string> props, 
                                I_ParticipantDBRepository participantRepo,
                                I_EventDBRepository eventRepo) 
        {
            log.Info($"{nameof(OfficeDBRepository)} constructed.");
            Props = props;
            participantRepository = participantRepo;
            eventRepository = eventRepo;
        }

        public IEnumerable<Participant> findParticipantsByEvent(long eventId)
        {
            log.Info($"Finding participants for event: {eventId}");
            var connection = DBUtils.GetConnection(Props);
            List<Participant> participants = new List<Participant>();
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT idParticipant, name, age FROM Offices WHERE idEvent = @eventId";
                    var param = command.CreateParameter();
                    param.ParameterName = "@eventId";
                    param.Value = eventId;
                    command.Parameters.Add(param);

                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            long id = reader.GetInt64(0);
                            string name = reader.GetString(1);
                            int age = reader.GetInt32(2);
                            Participant participant = new Participant(name, age) { Id = id };
                            participants.Add(participant);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error finding participants by event", ex);
                throw new EntityRepoException(ex);
            }
            
            return participants;
        }

        public void deleteByIDs(long participantID, long eventID)
        {
            log.Info($"Deleting task for participantID={participantID}, eventID={eventID}");
            var connection = DBUtils.GetConnection(Props);

            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "DELETE FROM Offices WHERE idParticipant = @idParticipant AND idEvent = @idEvent";

                var param1 = command.CreateParameter();
                param1.ParameterName = "@idParticipant";
                param1.Value = participantID;
                command.Parameters.Add(param1);

                var param2 = command.CreateParameter();
                param2.ParameterName = "@idEvent";
                param2.Value = eventID;
                command.Parameters.Add(param2);

                command.ExecuteNonQuery();
                log.Info($"Deleted task for participantID={participantID}");
            }
            catch (Exception ex)
            {
                log.Error($"Error deleting task for participantID={participantID}", ex);
                throw new EntityRepoException("Error deleting task");
            }
        }


        public int countEventsForParticipant(long participantId)
        {
            log.Info($"Counting events for participant: {participantId}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT COUNT(*) FROM Offices WHERE idParticipant = @participantId";
                    var param = command.CreateParameter();
                    param.ParameterName = "@participantId";
                    param.Value = participantId;
                    command.Parameters.Add(param);

                    var result = command.ExecuteScalar();
                    return Convert.ToInt32(result);
                }
            }
            catch (Exception ex)
            {
                log.Error("Error counting events for participant", ex);
                throw new EntityRepoException(ex);
            }
        }
        
        public void Add(Office entity)
        {
            log.Info($"Adding office: {entity}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = @"INSERT INTO Offices(idEvent, style, distance, idParticipant, name, age) 
                                          VALUES (@idEvent, @style, @distance, @idParticipant, @name, @age)";
                    
                    var idEventParam = command.CreateParameter();
                    idEventParam.ParameterName = "@idEvent";
                    idEventParam.Value = entity.Event.Id;
                    command.Parameters.Add(idEventParam);
                    
                    var styleParam = command.CreateParameter();
                    styleParam.ParameterName = "@style";
                    styleParam.Value = entity.Event.Style;
                    command.Parameters.Add(styleParam);
                    
                    var distanceParam = command.CreateParameter();
                    distanceParam.ParameterName = "@distance";
                    distanceParam.Value = entity.Event.Distance;
                    command.Parameters.Add(distanceParam);
                    
                    var idParticipantParam = command.CreateParameter();
                    idParticipantParam.ParameterName = "@idParticipant";
                    idParticipantParam.Value = entity.Participant.Id;
                    command.Parameters.Add(idParticipantParam);
                    
                    var nameParam = command.CreateParameter();
                    nameParam.ParameterName = "@name";
                    nameParam.Value = entity.Participant.Name;
                    command.Parameters.Add(nameParam);
                    
                    var ageParam = command.CreateParameter();
                    ageParam.ParameterName = "@age";
                    ageParam.Value = entity.Participant.Age;
                    command.Parameters.Add(ageParam);
                    
                    command.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                log.Error("Error adding office", ex);
                throw new EntityRepoException(ex);
            }
        }

        public void Remove(long id)
        {
            log.Info($"Removing office: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "DELETE FROM Offices WHERE id = @id";
                    var param = command.CreateParameter();
                    param.ParameterName = "@id";
                    param.Value = id;
                    command.Parameters.Add(param);
                    
                    command.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                log.Error("Error removing office", ex);
                throw new EntityRepoException(ex);
            }
        }

        public void Update(long id, Office entity)
        {
            log.Info($"Updating office: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = @"UPDATE Offices SET 
                                          idEvent = @idEvent, 
                                          style = @style, 
                                          distance = @distance, 
                                          idParticipant = @idParticipant, 
                                          name = @name, 
                                          age = @age 
                                          WHERE id = @id";
                    
                    var idEventParam = command.CreateParameter();
                    idEventParam.ParameterName = "@idEvent";
                    idEventParam.Value = entity.Event.Id;
                    command.Parameters.Add(idEventParam);
                    
                    var styleParam = command.CreateParameter();
                    styleParam.ParameterName = "@style";
                    styleParam.Value = entity.Event.Style;
                    command.Parameters.Add(styleParam);
                    
                    var distanceParam = command.CreateParameter();
                    distanceParam.ParameterName = "@distance";
                    styleParam.Value = entity.Event.Distance;
                    command.Parameters.Add(distanceParam);
                    
                    var idParticipantParam = command.CreateParameter();
                    idParticipantParam.ParameterName = "@idParticipant";
                    idParticipantParam.Value = entity.Participant.Id;
                    command.Parameters.Add(idParticipantParam);
                    
                    var nameParam = command.CreateParameter();
                    nameParam.ParameterName = "@name";
                    nameParam.Value = entity.Participant.Name;
                    command.Parameters.Add(nameParam);
                    
                    var ageParam = command.CreateParameter();
                    ageParam.ParameterName = "@age";
                    ageParam.Value = entity.Participant.Age;
                    command.Parameters.Add(ageParam);
                    
                    command.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                log.Error("Error updating office", ex);
                throw new EntityRepoException(ex);
            }
        }

        public Office findById(long id)
        {
            log.Info($"Finding office by ID: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Offices WHERE id = @id";
                    var param = command.CreateParameter();
                    param.ParameterName = "@id";
                    param.Value = id;
                    command.Parameters.Add(param);

                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            return Extract(reader);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error finding office by ID", ex);
                throw new EntityRepoException(ex);
            }

            return null;
        }

        public IEnumerable<Office> getAll()
        {
            log.Info("Getting all offices");
            var connection = DBUtils.GetConnection(Props);
            List<Office> offices = new List<Office>();
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Offices";
                    
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            offices.Add(Extract(reader));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error getting all offices", ex);
                throw new EntityRepoException(ex);
            }
            
            return offices;
        }

        public IEnumerable<Office> getEntriesByEvent(long eventId)
        {
            log.Info($"Getting offices for event: {eventId}");
            var connection = DBUtils.GetConnection(Props);
            List<Office> offices = new List<Office>();
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Offices WHERE idEvent = @eventId";
                    var param = command.CreateParameter();
                    param.ParameterName = "@eventId";
                    param.Value = eventId;
                    command.Parameters.Add(param);

                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            offices.Add(Extract(reader));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error getting offices by event", ex);
                throw new EntityRepoException(ex);
            }
            
            return offices;
        }
        

        private Office Extract(IDataReader reader)
        {
            long eventId = reader.GetInt64(reader.GetOrdinal("idEvent"));
            string style = reader.GetString(reader.GetOrdinal("style"));
            int distance = reader.GetInt32(reader.GetOrdinal("distance"));
            Event evt = new Event(style, distance) { Id = eventId };

            long participantId = reader.GetInt64(reader.GetOrdinal("idParticipant"));
            string name = reader.GetString(reader.GetOrdinal("name"));
            int age = reader.GetInt32(reader.GetOrdinal("age"));
            Participant participant = new Participant(name, age) { Id = participantId };

            return new Office(participant, evt);
        }
    }
}